### ISSUE TITLE
After upgrade of k3os to v0.9.0, boot fails with kernel panic - attempted to kill init

### ISSUE THREAD

**Issue by sgielen** [sgielen/picl-k3os-image-generator#4]
I was running k3os v0.6.0 built using the generator on a Raspberry Pi 3 model B.

After a k3os-upgrade-rootfs to k3os v0.9.0, boot failed with the following messages on the logs:

```
NAME:
   k3os - Booting to k3s so you don't have to

USAGE:
   k3os [global options] command [command options] [arguments...]

VERSION:
   v0.9.0

COMMANDS:
   rc           early phase "run commands" / "run control" 
   config, cfg  configure k3OS
   install      install k3OS
   upgrade      perform upgrades
   help, h      Shows a list of commands or help for one command

GLOBAL OPTIONS:
   --debug        Turn on debug logs [$K3OS_DEBUG]
   -help, -h      show help
   --version, -v  print the version
[    2.230329] Kernel panic - not syncing: Attempted to kill init! exitcode=0x00000000
```

**Comment by sgielen** [comment]
This looks like k3os requires an argument to run, now, while earlier it did not.

The init script just runs `/sbin/k3os` -- it looks like this needs to be `/sbin/k3os rc` now. Will attempt a fix.

**Comment by sgielen** [comment]
I changed it to `/sbin/k3os rc`. Now, the boot fails with these messages:

```
1970/01/01 00:00:02 error making directory /run: mkdir /run: read-only file system
1970/01/01 00:00:02 error mounting tmpfs to /run: no such file or directory
1970/01/01 00:00:02 error making directory /var/cache: mkdir /var/cache: read-only file system
[more such errors]
1970/01/01 00:00:02 error mounting dev to /dev: device or resource busy
1970/01/01 00:00:02 Failed to run /usr/sbin/mdev -s: fork/exec /usr/sbin/mdev: no such file or directory
1970/01/01 00:00:02 Failed to run hwclock: fork/exec /sbin/hwclock: no such file or directory
```

**Comment by sgielen** [comment]
Trying with this as init:

```
#!/bin/sh
modprobe squashfs
mount -o remount,rw /
if [ ! -f "/usr/sbin/mdev" ]; then
  echo "#!/bin/sh" >/usr/sbin/mdev
  echo "exit 0" >>/usr/sbin/mdev
  chmod +x /usr/sbin/mdev
fi
if [ ! -f "/sbin/hwclock" ]; then
  echo "#!/bin/sh" >/sbin/hwclock
  echo "exit 0" >>/sbin/hwclock
  chmod +x /sbin/hwclock
fi
exec /sbin/k3os rc
```

**Comment by sgielen** [comment]
Seems to be this code: https://github.com/rancher/k3os/blob/master/pkg/cli/rc/rc.go#L173

**Comment by sgielen** [comment]
Apparently, since that commit, `/sbin/k3os` cares whether it is being called with `argv[0]` set to `init` or `/sbin/init`. I should use `exec -a /sbin/init /sbin/k3os` to get k3os to boot in init mode. However, it looks like k3os actually calls back into `/sbin/init`, expecting to find itself, but instead finding our wrapper. When I add an echo at the start of `/sbin/init`, I get this kind of output:

```
Starting as PID 1.
[    1.849760] EXT4-fs (mmcblk0p2): re-mounted. Opts: (null)
Exec k3os...
[   12.366798] EXT4-fs (mmcblk0p2): re-mounted. Opts: (null)
Starting as PID 1.
Exec k3os...
[   22.836007] EXT4-fs (mmcblk0p2): re-mounted. Opts: (null)
Starting as PID 1.
```

It looks like k3os (when run as init) starts by remounting root, then exec'ing /sbin/init again?

**Comment by sgielen** [comment]
I wonder at this point, though, whether it wouldn't be easier to use `init=/sbin/my-init` instead of replacing /sbin/init. I might be able to do this for new images, but it might be difficult to do this for existing ones on both supported platforms.

**Comment by sgielen** [comment]
The first time our `init` runs its environment is:

```
SHLVL=1
HOME=/
TERM=linux
PATH=/sbin:/usr/sbin:/bin:/usr/bin
PWD=/
```

Subsequent times, the following variables are added/changed:

```
SHLVL=2
ENTER_DEVICE=/dev/loop0
ENTER_DATA=./k3os/data
ENTER_ROOT=/proc/self/exe
```

So I can use those to use `exec -a enter-chroot` instead of `exec -a /sbin/init`. It's a hack on top of a hack, but allright.

**Comment by sgielen** [comment]
Aaand... Checking if `$ENTER_ROOT` is nonempty, then calling with `-a enter-root` instead of `-a /sbin/init` got k3os to boot. Will update this task tomorrow with a proper fix.

**Comment by sgielen** [comment]
So setting init to something like this works:
```
#!/bin/sh
modprobe squashfs
if [ -z "$ENTER_ROOT" ]; then
  exec -a /sbin/init /sbin/k3os
else
  exec -a enter-root /sbin/k3os
fi
```

But, I don't really like it and it will likely break in future versions of k3os. So, instead, I revisited my earlier comment about setting `init=` in the cmdline and realized that I already do this for `init.resizefs`. So, I'll go with reinstalling `/sbin/init` as a symlink to `k3os`, but set a cmdline to `init=/sbin/init.preinit` and make that a tool that calls `/sbin/init`, unless we override it at runtime to run something else:

```name=/sbin/init.preinit
#!/bin/sh

set -e

# sleep a second for devices to settle
sleep 1

# if an alternative init exists, exec it first. This is useful for future
# updates, as the file you're reading right now is inaccessible after k3os
# boots.
if [ -x "/k3os/system/init.preinit" ]; then
       exec /k3os/system/init.preinit
fi

modprobe squashfs
exec /sbin/init
```

With this, we keep a `modprobe squashfs` (which k3os actually performs itself as well according to the source code, but it doesn't seem to work for me?) and even give ourselves upgrade capability if we need it in the future.

**Comment by sgielen** [comment]
For new images, I'll push a new commit momentarily. If you want to upgrade existing images, then make these changes *before* upgrading to k3os v0.9.0, by mounting the SD card to a compatible operating system. (If you've already upgraded to k3os v0.9.0 and found this issue, these changes should also fix your boot.)

For the orangepipc2 image:
- Mount the only partition on the SD card to a directory called 'root'
- In `root/boot/env.txt`, add to the end of the existing line ` init=/sbin/init.preinit` (there must be a space before `init=`).
For the raspberrypi image:
- Mount the first partition on the SD card to a directory called 'boot' and the second to 'root'.
- In `boot/cmdline.txt`, add to the end of the existing line ` init=/sbin/init.preinit` (there must be a space before `init=`).
Then, on both images:
- Remove the file at `root/sbin/init`. Then, create a new symlink: `ln -s k3os root/sbin/init`.
- Create a file at `root/sbin/init.preinit` with the following contents.
- `chmod +x root/sbin/init.preinit`.
```
#!/bin/sh

set -e

# sleep a second for devices to settle
sleep 1

# if an alternative init exists, exec it first. This is useful for future
# updates, as the file you're reading right now is inaccessible after k3os
# boots.
if [ -x "/k3os/system/init.preinit" ]; then
       exec /k3os/system/init.preinit
fi

modprobe squashfs
exec /sbin/init
```
- Then, cleanly unmount all partitions and try to boot. This should come up normally (even on earlier versions of k3os), after which you can perform `k3os-upgrade-rootfs`.

**Comment by sgielen** [comment]
For new images, the issue is resolved by commit 56708943283e06f02d84b48276bec2dfd94e4078. For existing ones, see the previous coment.
